sphinx_needs/lsp/esbonio.py

```diff
@@ -0,0 +1,385 @@
+def complete_need_link(
+    ls: NeedlsFeatures, params: CompletionContext, lines: List[str], line: str, word: str
+) -> List[CompletionItem]:
+    # specify the need type, e.g.,
+    # ->req
+    if word.count(">") == 1:
+        return [CompletionItem(label=need_type, detail="need type") for need_type in ls.needs_store.types]
+
+    word_parts = word.split(">")
+
+    # specify doc in which need is specified, e.g.,
+    # ->req>fusion/index.rst
+    if word.count(">") == 2:
+        requested_type = word_parts[1]  # e.g., req, test, ...
+        if requested_type in ls.needs_store.types:
+            return doc_completion_items(ls, ls.needs_store.docs_per_type[requested_type], word_parts[2])
+
+    # specify the exact need, e.g.,
+    # ->req>fusion/index.rst>REQ_001
+    if word.count(">") == 3:
+        requested_type = word_parts[1]  # e.g., req, test, ...
+        requested_doc = word_parts[2]  # [0:-4]  # without `.rst` file extension
+        if requested_doc in ls.needs_store.needs_per_doc:
+            substitution = word[word.find("->") :]
+            start_char = line.find(substitution)
+            line_number = params.position.line
+            return [
+                CompletionItem(
+                    label=need["id"],
+                    insert_text=need["id"],
+                    documentation=need["description"],
+                    detail=need["title"],
+                    additional_text_edits=[
+                        TextEdit(
+                            range=Range(
+                                start=Position(line=line_number, character=start_char),
+                                end=Position(
+                                    line=line_number,
+                                    character=start_char + len(substitution),
+                                ),
+                            ),
+                            new_text="",
+                        )
+                    ],
+                )
+                for need in ls.needs_store.needs_per_doc[requested_doc]
+                if need["type"] == requested_type
+            ]
+
+    return []
+
+
+def generate_hash(user_name: str, doc_uri: str, need_prefix: str, line_number: int) -> str:
+    salt = os.urandom(blake2b.SALT_SIZE)  # pylint: disable=no-member
+    return blake2b(
+        f"{user_name}{doc_uri}{need_prefix}{line_number}".encode(),
+        digest_size=4,
+        salt=salt,
+    ).hexdigest()
+
+
+def generate_need_id(
+    ls: NeedlsFeatures, params: CompletionContext, lines: List[str], word: str, need_type: Optional[str] = None
+) -> str:
+    """Generate a need ID including hash suffix."""
+
+    user_name = getpass.getuser()
+    doc_uri = params.doc.uri
+    line_number = params.position.line
+
+    if not need_type:
+        try:
+            match = re.search(".. ([a-z]+)::", lines[line_number - 1])
+            need_type = match.group(1)  # type: ignore[union-attr]
+        except AttributeError:
+            return "ID"
```

Comment:
perhaps this is a better construction-

```suggestion
        match = re.search(".. ([a-z]+)::", lines[line_number - 1])
        if match:
            need_type = match.group(1)
        else:
            return "ID"
```